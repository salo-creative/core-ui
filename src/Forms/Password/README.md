# Password

This is a compound component designed to simplify password captures within applications. The component has a locked set of regex rules to enforce strong passwords as well as optional password confirmation field.

*n.b.* THIS SHOULD ONLY BE USED WHEN CAPTURING PASSWORDS. Do not use this for cases where users are entering their password to login etc.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Password from '@salo/core-ui/Forms/Password';
```

Implement as follows

```javascript
<Password
  name='field_name'
  Input={ Input } // Pass input field that must be used
  value={ this.state.Password }
  onChange={ Password => this.setState({  Password }) }
/>
```