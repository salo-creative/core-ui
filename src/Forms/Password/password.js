import React from 'react';
import PropTypes from 'prop-types';
import { string } from 'yup';

// COMPONENTS & STYLES
import PasswordHelper from './password.helper';

// HELPERS & CONSTANTS
import { colours } from '../../helpers/colours';
import { passwordRegex } from '../../helpers/form';

const Password = (props) => {
  const {
    background,
    border,
    borderRadius,
    className,
    confirm,
    disabled,
    error,
    fontSize,
    margin,
    name,
    onChange,
    padding,
    required,
    showHelper,
    size,
    value,
    Input
  } = props;

  // Hold state of individual inputs within component
  const [state, setState] = React.useState({
    password: value,
    password_confirm: value,
    password_confirm_touched: false,
    matches: false
  });

  // On field blur we emit these back on the on change function to the
  // parent if fields match or not using the confirm functionality
  const handleBlur = () => {
    if (state.matches || !confirm) {
      onChange(state.password);
    }
  };

  // This is needed to trigger field validation when return is pressed to submit
  const handleKeyUp = ({ e }) => {
    if (e.keyCode === 13 && (state.matches || !confirm)) {
      handleBlur();
    }
  };

  const handleChange = ({ key, val }) => {
    // Check if passwords match
    let matches = false;
    let { password_confirm_touched } = state;
    if (key === 'password_confirm') {
      matches = val === state.password;
      password_confirm_touched = true;
    } else {
      matches = val === state.password_confirm;
    }
    setState({
      ...state,
      [key]: val,
      matches,
      password_confirm_touched
    });
  };

  const standardProps = {
    background,
    border,
    borderRadius,
    disabled,
    fontSize,
    margin,
    padding,
    size,
    onBlur: handleBlur,
    onKeyUp: handleKeyUp
  };

  return (
    <div className={ `password__wrapper ${ className }` }>
      <Input
        { ...standardProps }
        error={ error && !string().matches(new RegExp(passwordRegex)).required().isValidSync(state.password) }
        errorMessage='Your password is not valid'
        key={ `password_${ name }` }
        label='Password'
        name={ `password_${ name }` }
        placeholder='Enter a password'
        required={ required }
        type='password'
        value={ state.password }
        onChange={ ({ value: val }) => handleChange({ key: 'password', val }) }
      />
      { showHelper && <PasswordHelper password={ state.password } /> }
      { confirm && (
        <Input
          { ...standardProps }
          error={ !state.matches && (state.password_confirm_touched || error) }
          errorMessage='Your passwords do not match'
          key={ `password_confirm_${ name }` }
          label='Re-enter password'
          name={ `password_confirm_${ name }` }
          placeholder='Confirm password'
          required={ required }
          type='password'
          value={ state.password_confirm }
          onChange={ ({ value: val }) => handleChange({ key: 'password_confirm', val }) }
        />
      ) }
    </div>
  );
};

Password.defaultProps = {
  background: colours.paleGrey,
  border: '1px solid',
  borderRadius: '0.4rem',
  className: '',
  confirm: true,
  disabled: false,
  error: false,
  fontSize: '1.4rem',
  Input: null,
  margin: '0 0 2rem',
  onChange: () => null,
  padding: '0 1rem',
  required: false,
  showHelper: true,
  size: 'M',
  value: ''
};

Password.propTypes = {
  background: PropTypes.string,
  border: PropTypes.string,
  borderRadius: PropTypes.string,
  className: PropTypes.string,
  confirm: PropTypes.bool,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  fontSize: PropTypes.string,
  Input: PropTypes.func,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  padding: PropTypes.string,
  required: PropTypes.bool,
  showHelper: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  value: PropTypes.string
};

export default Password;