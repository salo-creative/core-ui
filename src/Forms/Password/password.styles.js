import styled from 'styled-components';

export const HelperWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: -1rem 0 3rem;
`;

export const ValidationLabel = styled.p`
  font-size: 1.2rem;
  margin: 0 0 1rem;
`;

export const ValidationRules = styled.span`
  display: flex;
  height: 2rem;
  font-size: 1.2rem;
  align-items: center;
  &.valid {
    color: ${ ({ theme }) => theme.primary };
    svg path {
      fill: ${ ({ theme }) => theme.primary };
    }
  }
`;