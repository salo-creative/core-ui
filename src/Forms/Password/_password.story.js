import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Password, Input } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Forms | Password', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['password'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: '' });
    // KNOBS
    const padding = text('Padding', '0 1rem');
    const border = text('Border', '1px solid');
    const borderRadius = text('Border radius', '0.4rem');
    const confirm = boolean('Show confirm', true);
    const fontSize = text('Font size', '1.4rem');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const size = select('size', ['L', 'M'], 'M');
    return (
      <RenderWithProps store={ store }>
        <Password
          border={ border }
          borderRadius={ borderRadius }
          confirm={ confirm }
          disabled={ disabled }
          error={ error }
          fontSize={ fontSize }
          Input={ Input }
          name='story'
          onChange={ (password) => console.log(password) }
          padding={ padding }
          required={ required }
          size={ size }
          value={ store.value }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);