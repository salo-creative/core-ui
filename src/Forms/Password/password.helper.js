import React from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';

// COMPONENTS & STYLES
import { HelperWrapper, ValidationLabel, ValidationRules } from './password.styles';

const Helper = (props) => {
  const { password } = props;

  const hasLength = password.length >= 8;
  const hasUppercase = new RegExp(/\w*[A-Z]\w*/).test(password);
  const hasLowercase = new RegExp(/\w*[a-z]\w*/).test(password);
  const hasNumber = new RegExp(/\w*[0-9]\w*/).test(password);
  const hasSpecial = new RegExp(/\w*[@$!%*?&£#]\w*/).test(password);

  const iconProps = {
    margin: '0 0.5rem 0 0',
    size: 18
  };

  return (
    <HelperWrapper>
      <ValidationLabel>Your password must contain the following:</ValidationLabel>
      <ValidationRules className={ hasLength ? 'valid' : '' }>
        <Icon
          { ...iconProps }
          icon={ hasLength ? 'tick' : 'alert' }
        />
            A minimum of 8 characters
      </ValidationRules>
      <ValidationRules className={ hasUppercase ? 'valid' : '' }>
        <Icon
          { ...iconProps }
          icon={ hasUppercase ? 'tick' : 'alert' }
        />
            Uppercase letter (A-Z)
      </ValidationRules>
      <ValidationRules className={ hasLowercase ? 'valid' : '' }>
        <Icon
          { ...iconProps }
          icon={ hasLowercase ? 'tick' : 'alert' }
        />
            Lowercase letter (a-z)
      </ValidationRules>
      <ValidationRules className={ hasNumber ? 'valid' : '' }>
        <Icon
          { ...iconProps }
          icon={ hasNumber ? 'tick' : 'alert' }
        />
            Number (0-9)
      </ValidationRules>
      <ValidationRules className={ hasSpecial ? 'valid' : '' }>
        <Icon
          { ...iconProps }
          icon={ hasSpecial ? 'tick' : 'alert' }
        />
            Special character (@$!%*?&£#)
      </ValidationRules>
    </HelperWrapper>
  );
};

Helper.defaultProps = { password: '' };

Helper.propTypes = { password: PropTypes.string };

export default Helper;