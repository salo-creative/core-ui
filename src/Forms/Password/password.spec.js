import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Password from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Password />
  );
  expect(container.firstChild).toMatchSnapshot();
});