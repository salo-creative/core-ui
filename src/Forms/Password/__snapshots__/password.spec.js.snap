// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c7 {
  height: 18px;
  width: 18px;
  margin: 0 0.5rem 0 0;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: -1rem 0 3rem;
}

.c5 {
  font-size: 1.2rem;
  margin: 0 0 1rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2rem;
  font-size: 1.2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6.valid {
  color: #00ACED;
}

.c6.valid svg path {
  fill: #00ACED;
}

.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 0 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  position: relative;
}

.c3 {
  margin: 0;
  padding: 0 1rem;
  height: 4rem;
  background-color: #fafcff;
  border: 1px solid;
  border-color: rgba(0,172,237,0.5);
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
  width: 100%;
  border-radius: 0.4rem;
  font-size: 1.4rem;
}

.c3:focus,
.c3:active {
  border-color: #00ACED;
}

.c3[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

.c1 {
  font-size: 1.4rem;
  margin: 0 0 0.5rem;
  padding: 0;
  color: inherit;
  word-break: break-word;
}

<div
  class="password__wrapper "
>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="password_undefined"
    >
      Password
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="password_undefined"
        name="password_undefined"
        placeholder="Enter a password"
        type="password"
        value=""
      />
    </div>
  </div>
  <div
    class="c4"
  >
    <p
      class="c5"
    >
      Your password must contain the following:
    </p>
    <span
      class="c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        focusable="false"
        size="18"
        viewBox="0 0 24 24"
      >
        <path
          d="M13,13H11V7H13M13,17H11V15H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
          fill="#00222b"
        />
      </svg>
      A minimum of 8 characters
    </span>
    <span
      class="c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        focusable="false"
        size="18"
        viewBox="0 0 24 24"
      >
        <path
          d="M13,13H11V7H13M13,17H11V15H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
          fill="#00222b"
        />
      </svg>
      Uppercase letter (A-Z)
    </span>
    <span
      class="c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        focusable="false"
        size="18"
        viewBox="0 0 24 24"
      >
        <path
          d="M13,13H11V7H13M13,17H11V15H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
          fill="#00222b"
        />
      </svg>
      Lowercase letter (a-z)
    </span>
    <span
      class="c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        focusable="false"
        size="18"
        viewBox="0 0 24 24"
      >
        <path
          d="M13,13H11V7H13M13,17H11V15H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
          fill="#00222b"
        />
      </svg>
      Number (0-9)
    </span>
    <span
      class="c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        focusable="false"
        size="18"
        viewBox="0 0 24 24"
      >
        <path
          d="M13,13H11V7H13M13,17H11V15H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
          fill="#00222b"
        />
      </svg>
      Special character (@$!%*?&Â£#)
    </span>
  </div>
  <div
    class="c0 salo-input "
  >
    <label
      class="c1 salo-input__label"
      for="password_confirm_undefined"
    >
      Re-enter password
    </label>
    <div
      class="c2 salo-input__wrapper"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="c3 salo-input__field"
        font-size="1.4rem"
        height="4rem"
        id="password_confirm_undefined"
        name="password_confirm_undefined"
        placeholder="Confirm password"
        type="password"
        value=""
      />
    </div>
  </div>
</div>
`;
